src/lib.rs

```diff
@@ -336,18 +339,63 @@ impl Config {
         assert!(output.ends_with(".a"));
         let lib_name = &output[3..output.len() - 2];
         let dst = self.get_out_dir();
+        let msvc = self.get_target().contains("msvc");
+        let compiler = self.get_compiler();
 
         let mut objects = Vec::new();
         let mut src_dst = Vec::new();
+        let mut dependencies = HashMap::new();
         for file in self.files.iter() {
             let obj = dst.join(file).with_extension("o");
             let obj = if !obj.starts_with(&dst) {
                 dst.join(obj.file_name().unwrap())
             } else {
                 obj
             };
+
+            let mut input_mtime = get_mtime(&file);
+            let output_mtime = get_mtime(&obj);
+            // If input file is older than output object, check its dependencies
+            if input_mtime < output_mtime {
+                let mut cmd = compiler.to_command();
+                let files: Vec<PathBuf>;
+                cmd.arg(&file);
+                if msvc {
+                    cmd.arg("/showIncludes").arg("/E");
+                    files = cmd.output()
+                        .map(|o| String::from_utf8(o.stderr))
+                        .unwrap().unwrap()
+                        .replace("Note: including file:", "")
+                        .lines().skip(1) // skip input files
+                        .map(|s| PathBuf::from(s.trim()))
+                        .collect();
+                } else {
+                    cmd.arg("-M").arg("-MG");
+                    files = cmd.output()
+                        .map(|o| String::from_utf8(o.stdout))
+                        .unwrap().unwrap()
+                        .replace(" \\", "")
```

Comment:
The output of `-M -MG` is a multiline `make` rule with backslashes for line continuation. This strips them out.

Hmm, looking at this now, I realise I'm not sure what the output would be with spaces in parent directories. I'll have to handle that.